Add Lutron Homeworks component

dubnom opened the pull request:
## Description:
Lutron made the Homeworks series 4 & 8 systems about 15 years ago.  These components interface with Homeworks.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7422

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

dubnom committed:
Added Lutron Homeworks components.

pvizeli reviewed (commented):
Use `helper/dispatcher` to send a signal from the main callback handler to platform. That help also to hold the worker time small for your library. Make sure that the platform first listens on dispatcher on `async_added_to_hass`. There exists some example of how that will work.

dubnom commented:
I'm used to Python 2.7, not all of the cool async features for 3.5+.  I'm getting there.  I also come from CVS source code control world, so moving to Git has been fun.  I had already switched my code to async_added_to_hass when I tested with a much larger set of devices.  Do you have a good example component that uses help/dispatcher?  Also, if I make these mods (or other mods), what is the correct way of getting these changes into the dev build?
Thanks - Mike

pvizeli commented:
Example:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/android_ip_webcam.py

So we support only python > 3.5.2 . You should use a python3 developer environment for Home Assistant.

dubnom commented:
Thanks for the examples. I'm fusing 3.6.5 in a virtual environment.

MartinHjelmare commented on homeassistant/components/binary_sensor/homeworks.py:
```diff
@@ -0,0 +1,86 @@
+    @property
+    def device_state_attributes(self):
+        """Return supported attributes."""
+        return {"Homeworks Address": self._addr,
```
State attribute keys should be lowercase snakecase.

MartinHjelmare commented on homeassistant/components/binary_sensor/homeworks.py:
```diff
@@ -0,0 +1,86 @@
+
+        old_state = self._state
+        if msg_type == HW_BUTTON_PRESSED and values[1] == self._num:
+            self.hass.bus.fire(EVENT_BUTTON_PRESSED,
```
Entities shouldn't fire events. Entities already spawn state changed events. Custom events can be fired from the component (homeworks) but not per entity.

MartinHjelmare commented on homeassistant/components/homeworks.py:
```diff
@@ -0,0 +1,92 @@
+https://home-assistant.io/components/homeworks/
+"""
+import logging
+import voluptuous as vol
```
See above.

MartinHjelmare commented on homeassistant/components/homeworks.py:
```diff
@@ -0,0 +1,92 @@
+
+        def __init__(self, host, port):
+            """Host and port of Lutron Homeworks controller."""
+            Homeworks.__init__(self, host, port, self.callback)
```
```suggestion
            super().__init__(host, port, self.callback)
```

MartinHjelmare commented on homeassistant/components/homeworks.py:
```diff
@@ -0,0 +1,92 @@
+            """Add a device to subscribe to events."""
+            if device.addr not in self._subscribers:
+                self._subscribers[device.addr] = []
+            self._subscribers[device.addr].append(device)
```
Don't store the entity in this shared container. Use our dispatch helper instead to signal the entity from the component. Connect the callback to the correct signal in the entity coroutine method `async_added_to_hass`. An entity specific signal can be created using a constant part and an entity or device specific part, eg the `entity.addr`.
```py
ENTITY_SIGNAL = 'homeworks_entity_{}'
...
signal = ENTITY_SIGNAL.format(addr)
```

MartinHjelmare commented on homeassistant/components/homeworks.py:
```diff
@@ -0,0 +1,92 @@
+            for sub in self._subscribers.get(addr, []):
+                _LOGGER.debug("_callback: %s", sub)
+                if sub.callback(msg_type, values):
+                    sub.schedule_update_ha_state()
```
We aren't allowed to access entity methods outside the platforms. This should be part of the update callback inside the entity.

MartinHjelmare commented on homeassistant/components/homeworks.py:
```diff
@@ -0,0 +1,92 @@
+        self._addr = addr
+        self._name = name
+        self._controller = controller
+        controller.register(self)
```
Callback registration should (dispatcher signal connection) should happen in `async_added_to_hass`.

MartinHjelmare commented on homeassistant/components/homeworks.py:
```diff
@@ -0,0 +1,92 @@
+        controller.register(self)
+
+    @property
+    def addr(self):
```
Remove the property and just make this a regular instance attribute.

MartinHjelmare commented on homeassistant/components/light/homeworks.py:
```diff
@@ -0,0 +1,105 @@
+"""Component for interfacing to Lutron Homeworks lights.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/homeworks/
```
Update url.

MartinHjelmare commented on homeassistant/components/light/homeworks.py:
```diff
@@ -0,0 +1,105 @@
+https://home-assistant.io/components/homeworks/
+"""
+import logging
+import voluptuous as vol
```
See above.

MartinHjelmare commented on homeassistant/components/light/homeworks.py:
```diff
@@ -0,0 +1,105 @@
+import homeassistant.helpers.config_validation as cv
+
+DEPENDENCIES = ['homeworks']
+REQUIREMENTS = ['pyhomeworks==0.0.1']
```
Remove this.

MartinHjelmare commented on homeassistant/components/light/homeworks.py:
```diff
@@ -0,0 +1,105 @@
+                             dimmer[CONF_NAME], dimmer[CONF_RATE])
+        devs.append(dev)
+    add_entities(devs, True)
+    return True
```
Remove this.

MartinHjelmare commented on homeassistant/components/light/homeworks.py:
```diff
@@ -0,0 +1,105 @@
+
+    def __init__(self, controller, addr, name, rate):
+        """Create device with Addr, name, and rate."""
+        HomeworksDevice.__init__(self, controller, addr, name)
```
```suggestion
        super().__init__(controller, addr, name)
```

MartinHjelmare commented on homeassistant/components/light/homeworks.py:
```diff
@@ -0,0 +1,105 @@
+        """Control the brightness."""
+        return self._level
+
+    @brightness.setter
```
Remove the setter. We don't use setters for entity state properties.

If the update callback is called within 2 seconds with feedback after calling the service in home assistant, we can let the callback update the brightness level.

MartinHjelmare commented on homeassistant/components/light/homeworks.py:
```diff
@@ -0,0 +1,105 @@
+    @property
+    def device_state_attributes(self):
+        """Supported attributes."""
+        return {'Homeworks Address': self._addr}
```
Use lowercase snakecase keys.

MartinHjelmare commented:
Make sure we explain in the docs how this integration is different than the existing lutron integrations:
https://www.home-assistant.io/components/lutron/
https://www.home-assistant.io/components/lutron_caseta/

dubnom commented:
I’ve made most of the changes you recommended to Lutron Homeworks, but I have a few questions before I work on the rest –

  1.  Lutron came up with a number of different systems they branded “Homeworks” and they each have totally different interfacing protocol.  Unfortunately, the Homeworks Series 4&8 systems have a form of autodiscovery but there is no documentation and it isn’t a supported API.  I can attempt to reverse engineer it, but that might take a long time (and still be incorrect).  This is why each device is added separately.  Knowing this, should I leave the way I was doing configuration, or should I still try to amass everything in one place?
  2.  In binary_sensor/homeworks.py you noted that entities shouldn’t fire events.  The keypad button is pressed and released within about 400 milliseconds.  I’m not sure that changing the state (and asynchronously notifying) will do the right thing such a small period of time.  Is there such a thing as a “momentary state change”?
  3.  I followed the suggestion of using ‘super()’ instead of explicit class names, but I’m not sure this is clearer when using multiple inheritance.  Thoughts?
  4.  I copied a bit from the Lutron implementation which didn’t use signals for the callbacks.  Do you have an implementation that you like that I should look at?
Thanks!

  *   Mike

From: Martin Hjelmare <notifications@github.com>
Sent: Tuesday, November 13, 2018 5:19 AM
To: home-assistant/home-assistant <home-assistant@noreply.github.com>
Cc: Michael Dubno <michael@dubno.com>; Author <author@noreply.github.com>
Subject: Re: [home-assistant/home-assistant] Added Lutron Homeworks components. (#18311)


@MartinHjelmare requested changes on this pull request.

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232965090&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=Gml4r1pDlLwG6MtZjXl5MMEWUkiseeo0ylwbZ4yNZWk&e=>:

> @@ -0,0 +1,86 @@

+"""Component for interfacing to Lutron Homeworks keypads.

+

+For more details about this component, please refer to the documentation at

+https://home-assistant.io/components/homeworks/<https://urldefense.proofpoint.com/v2/url?u=https-3A__home-2Dassistant.io_components_homeworks_&d=DwQFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=6_dxgzEFCM9CpB3fDbzZ9CvffXoIs8RAbU9lqE0cCfc&e=>

The url should point to the binary sensor docs page. See other platforms for correct format.

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232965301&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=fsF7JO2o5hzPFyXmHhTbxbiooA9q1tmAQapvA1tBwmc&e=>:

> @@ -0,0 +1,86 @@

+"""Component for interfacing to Lutron Homeworks keypads.

+

+For more details about this component, please refer to the documentation at

+https://home-assistant.io/components/homeworks/<https://urldefense.proofpoint.com/v2/url?u=https-3A__home-2Dassistant.io_components_homeworks_&d=DwQFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=6_dxgzEFCM9CpB3fDbzZ9CvffXoIs8RAbU9lqE0cCfc&e=>

+"""

+import logging

+import voluptuous as vol

Please add a blank line between standard library, 3rd party and homeassistant import groups.

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232965530&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=KDxznfXR2sRzPLEKgcS7pbvcjLj3u5oWqlnv_ys4-nM&e=>:

> +"""Component for interfacing to Lutron Homeworks keypads.

+

+For more details about this component, please refer to the documentation at

+https://home-assistant.io/components/homeworks/<https://urldefense.proofpoint.com/v2/url?u=https-3A__home-2Dassistant.io_components_homeworks_&d=DwQFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=6_dxgzEFCM9CpB3fDbzZ9CvffXoIs8RAbU9lqE0cCfc&e=>

+"""

+import logging

+import voluptuous as vol

+from homeassistant.components.binary_sensor import (

+    BinarySensorDevice, PLATFORM_SCHEMA)

+from homeassistant.components.homeworks import (

+    HomeworksDevice, HOMEWORKS_CONTROLLER)

+from homeassistant.const import CONF_NAME

+import homeassistant.helpers.config_validation as cv

+

+DEPENDENCIES = ['homeworks']

+REQUIREMENTS = ['pyhomeworks==0.0.1']

We can remove this. The component will require this and here in the platform we depend on the component.

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232966514&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=ucMIWYxQqz1rBTBEUH3OrK46Q3sJwb3HTmCQx-xg3lw&e=>:

> +REQUIREMENTS = ['pyhomeworks==0.0.1']

+

+_LOGGER = logging.getLogger(__name__)

+

+EVENT_BUTTON_PRESSED = 'button_pressed'

+CONF_KEYPADS = 'keypads'

+CONF_ADDR = 'addr'

+CONF_BUTTONS = 'buttons'

+

+BUTTON_SCHEMA = vol.Schema({cv.positive_int: cv.string})

+BUTTONS_SCHEMA = vol.Schema({

+    vol.Required(CONF_ADDR): cv.string,

+    vol.Required(CONF_NAME): cv.string,

+    vol.Required(CONF_BUTTONS): vol.All(cv.ensure_list, [BUTTON_SCHEMA])

+})

+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({

It's better to keep all the config under the component and load platforms via discovery. See the rainmachine component for an example.

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232967166&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=cPFF_jZUUjwyOJE9lSKBSbwAnOCin1MtM7Qnu3MbjZE&e=>:

> +BUTTON_SCHEMA = vol.Schema({cv.positive_int: cv.string})

+BUTTONS_SCHEMA = vol.Schema({

+    vol.Required(CONF_ADDR): cv.string,

+    vol.Required(CONF_NAME): cv.string,

+    vol.Required(CONF_BUTTONS): vol.All(cv.ensure_list, [BUTTON_SCHEMA])

+})

+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({

+    vol.Required(CONF_KEYPADS): vol.All(cv.ensure_list, [BUTTONS_SCHEMA])

+})

+

+

+def setup_platform(hass, config, add_entities, discover_info=None):

+    """Set up the Homeworks keypads."""

+    controller = hass.data[HOMEWORKS_CONTROLLER]

+    devs = []

+    for keypad in config.get(CONF_KEYPADS):

Use dict[key] for required config keys.

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232967616&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=izZIuX6-mkIYIsI7FiLS6vVxv7cjL98U8EX_cmRSIOg&e=>:

> +    vol.Required(CONF_KEYPADS): vol.All(cv.ensure_list, [BUTTONS_SCHEMA])

+})

+

+

+def setup_platform(hass, config, add_entities, discover_info=None):

+    """Set up the Homeworks keypads."""

+    controller = hass.data[HOMEWORKS_CONTROLLER]

+    devs = []

+    for keypad in config.get(CONF_KEYPADS):

+        name = keypad.get(CONF_NAME)

+        addr = keypad.get(CONF_ADDR)

+        buttons = keypad.get(CONF_BUTTONS)

+        for button in buttons:

+            # FIX: This should be done differently

+            for num, title in button.items():

+                devname = name + '_' + title

Use new style string formatting instead of string concatenation.
⬇️ Suggested change

-                devname = name + '_' + title

+                devname = '{}_{}'.format(name, title)

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232967902&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=0QG4G6PMZWUDf958hwRRwFLp7wpopdkV80CJV0jwJzU&e=>:

> +def setup_platform(hass, config, add_entities, discover_info=None):

+    """Set up the Homeworks keypads."""

+    controller = hass.data[HOMEWORKS_CONTROLLER]

+    devs = []

+    for keypad in config.get(CONF_KEYPADS):

+        name = keypad.get(CONF_NAME)

+        addr = keypad.get(CONF_ADDR)

+        buttons = keypad.get(CONF_BUTTONS)

+        for button in buttons:

+            # FIX: This should be done differently

+            for num, title in button.items():

+                devname = name + '_' + title

+                dev = HomeworksKeypad(controller, addr, num, devname)

+                devs.append(dev)

+    add_entities(devs, True)

+    return True

Nothing is checking this return value. We can remove the statement.

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232968265&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=8bTa4uPhPmf9ABC71vzaXd7XUzgxpNHGIrUNinBKhsg&e=>:

> +        for button in buttons:

+            # FIX: This should be done differently

+            for num, title in button.items():

+                devname = name + '_' + title

+                dev = HomeworksKeypad(controller, addr, num, devname)

+                devs.append(dev)

+    add_entities(devs, True)

+    return True

+

+

+class HomeworksKeypad(HomeworksDevice, BinarySensorDevice):

+    """Homeworks Keypad."""

+

+    def __init__(self, controller, addr, num, name):

+        """Create keypad with addr, num, and name."""

+        HomeworksDevice.__init__(self, controller, addr, name)
⬇️ Suggested change

-        HomeworksDevice.__init__(self, controller, addr, name)

+        super().__init__(controller, addr, name)

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232968425&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=OsBwAfSRX9NokGrsj_y3rVRnHI7RUU0g0Y09GX6bEKE&e=>:

> +

+    def __init__(self, controller, addr, num, name):

+        """Create keypad with addr, num, and name."""

+        HomeworksDevice.__init__(self, controller, addr, name)

+        self._num = num

+        self._state = None

+

+    @property

+    def is_on(self):

+        """Return state of the button."""

+        return self._state

+

+    @property

+    def device_state_attributes(self):

+        """Return supported attributes."""

+        return {"Homeworks Address": self._addr,

State attribute keys should be lowercase snakecase.

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232969087&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=irWaz8L32Vtc2rC-7rny73gC4Tsq71ioobZAQoGSiH4&e=>:

> +        return self._state

+

+    @property

+    def device_state_attributes(self):

+        """Return supported attributes."""

+        return {"Homeworks Address": self._addr,

+                "Button Number": self._num}

+

+    def callback(self, msg_type, values):

+        """Dispatch messages from the controller."""

+        from pyhomeworks.pyhomeworks import (

+            HW_BUTTON_PRESSED, HW_BUTTON_RELEASED)

+

+        old_state = self._state

+        if msg_type == HW_BUTTON_PRESSED and values[1] == self._num:

+            self.hass.bus.fire(EVENT_BUTTON_PRESSED,

Entities shouldn't fire events. Entities already spawn state changed events. Custom events can be fired from the component (homeworks) but not per entity.

________________________________

In homeassistant/components/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232969660&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=SYqDu4AhoCQP1NUz97MaXn6AO-4p_fljfojg-nk4SWU&e=>:

> @@ -0,0 +1,92 @@

+"""Component for interfacing to Lutron Homeworks Series 4 and 8 systems.

+

+For more details about this component, please refer to the documentation at

+https://home-assistant.io/components/homeworks/<https://urldefense.proofpoint.com/v2/url?u=https-3A__home-2Dassistant.io_components_homeworks_&d=DwQFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=6_dxgzEFCM9CpB3fDbzZ9CvffXoIs8RAbU9lqE0cCfc&e=>

+"""

+import logging

+import voluptuous as vol

See above.

________________________________

In homeassistant/components/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232969908&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=n4go4LGx_iHSpH-gvc6HC8Iy4-0cTvx4jT_vbDKXgeY&e=>:

> +        vol.Required(CONF_HOST): cv.string,

+        vol.Required(CONF_PORT): cv.port,

+    }),

+}, extra=vol.ALLOW_EXTRA)

+

+

+def setup(hass, base_config):

+    """Start Homeworks controller."""

+    from pyhomeworks.pyhomeworks import Homeworks

+

+    class HomeworksController(Homeworks):

+        """Interface between HASS and Homeworks controller."""

+

+        def __init__(self, host, port):

+            """Host and port of Lutron Homeworks controller."""

+            Homeworks.__init__(self, host, port, self.callback)
⬇️ Suggested change

-            Homeworks.__init__(self, host, port, self.callback)

+            super().__init__(host, port, self.callback)

________________________________

In homeassistant/components/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232971618&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=ooJUAKcuFSoOBN1QGvOm8ZSkZyZrChDO0ExEzi1GIks&e=>:

> +    """Start Homeworks controller."""

+    from pyhomeworks.pyhomeworks import Homeworks

+

+    class HomeworksController(Homeworks):

+        """Interface between HASS and Homeworks controller."""

+

+        def __init__(self, host, port):

+            """Host and port of Lutron Homeworks controller."""

+            Homeworks.__init__(self, host, port, self.callback)

+            self._subscribers = {}

+

+        def register(self, device):

+            """Add a device to subscribe to events."""

+            if device.addr not in self._subscribers:

+                self._subscribers[device.addr] = []

+            self._subscribers[device.addr].append(device)

Don't store the entity in this shared container. Use our dispatch helper instead to signal the entity from the component. Connect the callback to the correct signal in the entity coroutine method async_added_to_hass. An entity specific signal can be created using a constant part and an entity or device specific part, eg the entity.addr.

ENTITY_SIGNAL = 'homeworks_entity_{}'

...

signal = ENTITY_SIGNAL.format(addr)

________________________________

In homeassistant/components/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232972202&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=MQ5yjCD3u3M61HHyk3Oeoh8yBG-voRR-sm0yo3DgQtc&e=>:

> +            self._subscribers = {}

+

+        def register(self, device):

+            """Add a device to subscribe to events."""

+            if device.addr not in self._subscribers:

+                self._subscribers[device.addr] = []

+            self._subscribers[device.addr].append(device)

+

+        def callback(self, msg_type, values):

+            """Dispatch state changes."""

+            _LOGGER.debug('callback: %s, %s', msg_type, values)

+            addr = values[0]

+            for sub in self._subscribers.get(addr, []):

+                _LOGGER.debug("_callback: %s", sub)

+                if sub.callback(msg_type, values):

+                    sub.schedule_update_ha_state()

We aren't allowed to access entity methods outside the platforms. This should be part of the update callback inside the entity.

________________________________

In homeassistant/components/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232972599&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=p0YqQrSKpz_Qgng-0-_7Y0LqWpHhUlWqcV59I-GBgCw&e=>:

> +    def cleanup(event):

+        controller.close()

+

+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, cleanup)

+    return True

+

+

+class HomeworksDevice():

+    """Base class of a Homeworks device."""

+

+    def __init__(self, controller, addr, name):

+        """Controller, address, and name of the device."""

+        self._addr = addr

+        self._name = name

+        self._controller = controller

+        controller.register(self)

Callback registration should (dispatcher signal connection) should happen in async_added_to_hass.

________________________________

In homeassistant/components/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232972801&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=-VmjbZeAaRU-LfQo7TjgLC6t4Dzrscju2X3-QbTu7hU&e=>:

> +    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, cleanup)

+    return True

+

+

+class HomeworksDevice():

+    """Base class of a Homeworks device."""

+

+    def __init__(self, controller, addr, name):

+        """Controller, address, and name of the device."""

+        self._addr = addr

+        self._name = name

+        self._controller = controller

+        controller.register(self)

+

+    @property

+    def addr(self):

Remove the property and just make this a regular instance attribute.

________________________________

In homeassistant/components/light/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232972847&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=w_ifzOu2lvK4jBT_MlOvq5cAows9SqSpcZ1xb1mHF3g&e=>:

> @@ -0,0 +1,105 @@

+"""Component for interfacing to Lutron Homeworks lights.

+

+For more details about this component, please refer to the documentation at

+https://home-assistant.io/components/homeworks/<https://urldefense.proofpoint.com/v2/url?u=https-3A__home-2Dassistant.io_components_homeworks_&d=DwQFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=6_dxgzEFCM9CpB3fDbzZ9CvffXoIs8RAbU9lqE0cCfc&e=>

Update url.

________________________________

In homeassistant/components/light/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232972888&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=kQEMeEpwrof34w5P6S3UP_gc1Pk_KuokgB6BBieHsuc&e=>:

> @@ -0,0 +1,105 @@

+"""Component for interfacing to Lutron Homeworks lights.

+

+For more details about this component, please refer to the documentation at

+https://home-assistant.io/components/homeworks/<https://urldefense.proofpoint.com/v2/url?u=https-3A__home-2Dassistant.io_components_homeworks_&d=DwQFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=6_dxgzEFCM9CpB3fDbzZ9CvffXoIs8RAbU9lqE0cCfc&e=>

+"""

+import logging

+import voluptuous as vol

See above.

________________________________

In homeassistant/components/light/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232972971&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=3MwrOcIEkFah691Nde_MUdGPAse26mIP3a0ZUVrpGYU&e=>:

> +"""Component for interfacing to Lutron Homeworks lights.

+

+For more details about this component, please refer to the documentation at

+https://home-assistant.io/components/homeworks/<https://urldefense.proofpoint.com/v2/url?u=https-3A__home-2Dassistant.io_components_homeworks_&d=DwQFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=6_dxgzEFCM9CpB3fDbzZ9CvffXoIs8RAbU9lqE0cCfc&e=>

+"""

+import logging

+import voluptuous as vol

+from homeassistant.components.homeworks import (

+    HomeworksDevice, HOMEWORKS_CONTROLLER)

+from homeassistant.components.light import (

+    ATTR_BRIGHTNESS, SUPPORT_BRIGHTNESS, Light, PLATFORM_SCHEMA)

+from homeassistant.const import CONF_NAME

+import homeassistant.helpers.config_validation as cv

+

+DEPENDENCIES = ['homeworks']

+REQUIREMENTS = ['pyhomeworks==0.0.1']

Remove this.

________________________________

In homeassistant/components/light/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232973139&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=3od4RVd_osom88pnNszBZLIBTfobAw9VPXCbRol7CAM&e=>:

> +})

+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({

+    vol.Required(CONF_DIMMERS): vol.All(cv.ensure_list, [DIMMER_SCHEMA])

+})

+

+

+def setup_platform(hass, config, add_entities, discover_info=None):

+    """Set up the Homeworks lights."""

+    controller = hass.data[HOMEWORKS_CONTROLLER]

+    devs = []

+    for dimmer in config.get(CONF_DIMMERS):

+        dev = HomeworksLight(controller, dimmer[CONF_ADDR],

+                             dimmer[CONF_NAME], dimmer[CONF_RATE])

+        devs.append(dev)

+    add_entities(devs, True)

+    return True

Remove this.

________________________________

In homeassistant/components/light/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232973223&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=6NxHzse1dpH6MC0ubZcKZUZ9PhHGCZPSZ1FqsNDjaNE&e=>:

> +    controller = hass.data[HOMEWORKS_CONTROLLER]

+    devs = []

+    for dimmer in config.get(CONF_DIMMERS):

+        dev = HomeworksLight(controller, dimmer[CONF_ADDR],

+                             dimmer[CONF_NAME], dimmer[CONF_RATE])

+        devs.append(dev)

+    add_entities(devs, True)

+    return True

+

+

+class HomeworksLight(HomeworksDevice, Light):

+    """Homeworks Light."""

+

+    def __init__(self, controller, addr, name, rate):

+        """Create device with Addr, name, and rate."""

+        HomeworksDevice.__init__(self, controller, addr, name)
⬇️ Suggested change

-        HomeworksDevice.__init__(self, controller, addr, name)

+        super().__init__(controller, addr, name)

________________________________

In homeassistant/components/light/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232974126&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=1YExRYSiSfrBIVmKylab2T5AD7cNV6Q-_jyRrVOV6RA&e=>:

> +        """Turn on the light."""

+        if ATTR_BRIGHTNESS in kwargs:

+            self.brightness = kwargs[ATTR_BRIGHTNESS]

+        else:

+            self.brightness = 255

+

+    def turn_off(self, **kwargs):

+        """Turn off the light."""

+        self.brightness = 0

+

+    @property

+    def brightness(self):

+        """Control the brightness."""

+        return self._level

+

+    @brightness.setter

Remove the setter. We don't use setters for entity state properties.

If the update callback is called within 2 seconds with feedback after calling the service in home assistant, we can let the callback update the brightness level.

________________________________

In homeassistant/components/light/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr232974292&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=xCWiJfZhDcOy9O8M0NsV0DuSo4nfrrLcLQpJ9ATKRBo&e=>:

> +    @property

+    def brightness(self):

+        """Control the brightness."""

+        return self._level

+

+    @brightness.setter

+    def brightness(self, level):

+        self._controller.fade_dim(

+            float((level*100.)/255.), self._rate,

+            0, self._addr)

+        self._level = level

+

+    @property

+    def device_state_attributes(self):

+        """Supported attributes."""

+        return {'Homeworks Address': self._addr}

Use lowercase snakecase keys.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23pullrequestreview-2D174273907&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=D9ddU83Gqo2Yz05NIG786se1T70_Tv7Jj7Z4E_Mb3QA&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AQtHig0csJiwrsoh3ICmsMyYfAuwpM4Fks5uupyUgaJpZM4YS9US&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=rUnwZzP-KQwIeGMiR6pRL6w_qtwJpoOuoNpf0_KVaH8&s=spUxLIhbmOVGvCDETxwFNpUzsQac0MJEPhMXCyTPI2U&e=>.



________________________________

This email has been scanned for spam and viruses by Proofpoint Essentials. Click here<https://us3.proofpointessentials.com/index01.php?mod_id=11&mod_option=logitem&mail_id=1542104342-AEgvy_ZHj5FJ&r_address=michael%40dubno.com&report=1> to report this email as spam.

dubnom committed:
Made all requested changes other than config.

dubnom committed:
Removed commented out code.

MartinHjelmare commented:
1. Devices can still be configured individually from the component. Look at the octoprint component, or eg IHC component.
2. Look at the deconz component for how to fire events for eg buttons.
3. The `Entity` base class doesn't have an init method, so there's no confusion about multiple inheritance.
4. Use of dispatcher looks good after a quick glance. I'll look more asap.

dubnom committed:
Removed binary_sensor.
Implemented new signal/events for button presses.
Cleaned up some data passing.

dubnom commented:
I think I’ve done all the changes.  It makes more sense now.  I’ll update the docs once you approve the new config work.
Everything works nicely now.

The way buttons are implemented still seems a little weird too me.  In the case of the (actual) Homeworks system, buttons have press and release messages but they can also be “virtually pressed” by sending them a command.  I didn’t implement this feature.  This would look like a momentary switch of some sort.  By being an entity, it would also allow an easier history to be maintained (and displayed) of button presses.  Now my keypads are hidden and have no history.

Thanks for everything - Mike

From: Martin Hjelmare <notifications@github.com>
Sent: Saturday, December 15, 2018 10:55 AM
To: home-assistant/home-assistant <home-assistant@noreply.github.com>
Cc: Michael Dubno <michael@dubno.com>; Author <author@noreply.github.com>
Subject: Re: [home-assistant/home-assistant] Add Lutron Homeworks component (#18311)


@MartinHjelmare commented on this pull request.

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr241944133&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=f8U4mklHjajBzgo8zgQ_U_c3RhkJWneI5v7YBVxOOTc&s=2MwXZONw67aP2Jzxd4QRXVMzWtjR7CJDYwyOSwq38Jk&e=>:

> +        """Return state of the button."""

+        return self._state

+

+    @property

+    def device_state_attributes(self):

+        """Return state attributes."""

+        return {"HomeworksAddress": self._addr,

+                "ButtonNumber": self._num}

+

+    @callback

+    def _update_callback(self, msg_type, data):

+        """Dispatch messages from the controller."""

+        from pyhomeworks.pyhomeworks import (

+            HW_BUTTON_PRESSED, HW_BUTTON_RELEASED)

+

+        msg_type, values = data

We already have accesst to msg_type as an argument. The dispatch helper can send any number of arguments to the target _update_callback. So we can remove this line.

________________________________

In homeassistant/components/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr241944165&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=f8U4mklHjajBzgo8zgQ_U_c3RhkJWneI5v7YBVxOOTc&s=qPZ67CT-2dKXca7MwxuacJyAkddLO9DStPbdD9x6dk0&e=>:

> +    """Start Homeworks controller."""

+    from pyhomeworks.pyhomeworks import Homeworks

+

+    class HomeworksController(Homeworks):

+        """Interface between HASS and Homeworks controller."""

+

+        def __init__(self, host, port):

+            """Host and port of Lutron Homeworks controller."""

+            super().__init__(host, port, self.callback)

+

+        def callback(self, msg_type, values):

+            """Dispatch state changes."""

+            _LOGGER.debug('callback: %s, %s', msg_type, values)

+            addr = values[0]

+            signal = ENTITY_SIGNAL.format(addr)

+            dispatcher_send(hass, signal, (msg_type, values))

Send the arguments separately.
⬇️ Suggested change

-            dispatcher_send(hass, signal, (msg_type, values))

+            dispatcher_send(hass, signal, msg_type, values)

________________________________

In homeassistant/components/light/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr241944262&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=f8U4mklHjajBzgo8zgQ_U_c3RhkJWneI5v7YBVxOOTc&s=ruQkFIR-1uWwIjXb5tIkIvahJ5IMMq-WLlUHKQYvlWk&e=>:

> +        self._controller.fade_dim(

+            float((level*100.)/255.), self._rate,

+            0, self._addr)

+

+    @property

+    def device_state_attributes(self):

+        """Supported attributes."""

+        return {'HomeworksAddress': self._addr}

+

+    @property

+    def is_on(self):

+        """Is the light on/off."""

+        return self._level != 0

+

+    @callback

+    def _update_callback(self, data):
⬇️ Suggested change

-    def _update_callback(self, data):

+    def _update_callback(self, msg_type, values):

________________________________

In homeassistant/components/light/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr241944269&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=f8U4mklHjajBzgo8zgQ_U_c3RhkJWneI5v7YBVxOOTc&s=gpSOvJI757DoL2ZwlwOxKrmLOAeK5iBiCktktLgqMf0&e=>:

> +    def device_state_attributes(self):

+        """Supported attributes."""

+        return {'HomeworksAddress': self._addr}

+

+    @property

+    def is_on(self):

+        """Is the light on/off."""

+        return self._level != 0

+

+    @callback

+    def _update_callback(self, data):

+        """Process device specific messages."""

+        from pyhomeworks.pyhomeworks import HW_LIGHT_CHANGED

+

+        _LOGGER.debug('_update_callback %s', data)

+        msg_type, values = data

Remove this.

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr241944283&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=f8U4mklHjajBzgo8zgQ_U_c3RhkJWneI5v7YBVxOOTc&s=LhW9ep4nGJVCFPyjNrwT9sDgD7s6bAPnb683GKdcq1w&e=>:

> +        async_dispatcher_connect(

+            self.hass, signal, self._update_callback)

+

+    @property

+    def is_on(self):

+        """Return state of the button."""

+        return self._state

+

+    @property

+    def device_state_attributes(self):

+        """Return state attributes."""

+        return {"HomeworksAddress": self._addr,

+                "ButtonNumber": self._num}

+

+    @callback

+    def _update_callback(self, msg_type, data):
⬇️ Suggested change

-    def _update_callback(self, msg_type, data):

+    def _update_callback(self, msg_type, values):

________________________________

In homeassistant/components/binary_sensor/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr241944295&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=f8U4mklHjajBzgo8zgQ_U_c3RhkJWneI5v7YBVxOOTc&s=7IKIaSQWVDP6MqsYlahoLBuikUlry7v695dVhqojw6s&e=>:

> +

+    async def async_def_added_to_hass(self):

+        """Called when entity is added to hass."""

+        signal = ENTITY_SIGNAL.format(self._addr)

+        async_dispatcher_connect(

+            self.hass, signal, self._update_callback)

+

+    @property

+    def is_on(self):

+        """Return state of the button."""

+        return self._state

+

+    @property

+    def device_state_attributes(self):

+        """Return state attributes."""

+        return {"HomeworksAddress": self._addr,
⬇️ Suggested change

-        return {"HomeworksAddress": self._addr,

+        return {'homeworks_address': self._addr,

________________________________

In homeassistant/components/light/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr241944309&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=f8U4mklHjajBzgo8zgQ_U_c3RhkJWneI5v7YBVxOOTc&s=WHVllPf2zIeBQp9LPJ0yypisQ8oKJABZI6VHwuaz6V8&e=>:

> +

+    @property

+    def brightness(self):

+        """Control the brightness."""

+        return self._level

+

+    def _set_brightness(self, level):

+        """Send the brightness level to the device."""

+        self._controller.fade_dim(

+            float((level*100.)/255.), self._rate,

+            0, self._addr)

+

+    @property

+    def device_state_attributes(self):

+        """Supported attributes."""

+        return {'HomeworksAddress': self._addr}
⬇️ Suggested change

-        return {'HomeworksAddress': self._addr}

+        return {'homeworks_address': self._addr}

________________________________

In homeassistant/components/homeworks.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23discussion-5Fr241944335&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=f8U4mklHjajBzgo8zgQ_U_c3RhkJWneI5v7YBVxOOTc&s=5dEksjZhbTyB2gbzY4Sy6XPGTFno1KgkReivz6mqAOg&e=>:

> +    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, cleanup)

+    return True

+

+

+class HomeworksDevice():

+    """Base class of a Homeworks device."""

+

+    def __init__(self, controller, addr, name):

+        """Controller, address, and name of the device."""

+        self._addr = addr

+        self._name = name

+        self._controller = controller

+        controller.register(self)

+

+    @property

+    def addr(self):

Just remove this method. We don't seem to need it.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23pullrequestreview-2D185349092&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=f8U4mklHjajBzgo8zgQ_U_c3RhkJWneI5v7YBVxOOTc&s=kyTl2-kW3onhgt4fMIENWnJ4vBEyQV7wmidYGMWKI1Q&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AQtHilq3Zhsnz71xu39HmmOuQe-2DBfWoZks5u5RtcgaJpZM4YS9US&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=f8U4mklHjajBzgo8zgQ_U_c3RhkJWneI5v7YBVxOOTc&s=ohgHXckBHwDWG6hfLJPDLJ2xWJHNxxw94T5do3PMrdQ&e=>.



________________________________

This email has been scanned for spam and viruses by Proofpoint Essentials. Click here<https://us3.proofpointessentials.com/index01.php?mod_id=11&mod_option=logitem&mail_id=1544889316-nnYLiqUJ84oP&r_address=michael%40dubno.com&report=1> to report this email as spam.

dubnom committed:
Fixed minor formatting.

dubnom commented:
I think binary_sensor.octoprint’s documentation is incorrect and should look more like ‘sensor.octoprint’ (judging from the code).
Thanks - Mike

dubnom committed:
Got rid of unused config stuff.
Reordered imports.

dubnom committed:
Missed removing vol, and forgot an extra line.

MartinHjelmare commented on homeassistant/components/homeworks.py:
```diff
@@ -0,0 +1,151 @@
+            addr = values[0]
+            signal = ENTITY_SIGNAL.format(addr)
+            dispatcher_send(hass, signal, msg_type, values)
+            _LOGGER.debug('callback returned')
```
The callback will return before the signal target is executed. So this debug message doesn't add very much.

MartinHjelmare commented on homeassistant/components/light/homeworks.py:
```diff
@@ -0,0 +1,104 @@
+    @property
+    def device_state_attributes(self):
+        """Supported attributes."""
+        return {'homeworks_address': self._addr}
```
Is the address a unique id of the entity? Then we could implement the `unique_id` entity property which will activate the entity registry for the platform entities.

https://developers.home-assistant.io/docs/en/entity_registry_index.html#unique-id-requirements

MartinHjelmare commented on homeassistant/components/light/homeworks.py:
```diff
@@ -0,0 +1,104 @@
+            self._level = int((values[1] * 255.)/100.)
+            if self._level != 0:
+                self._prev_level = self._level
+            self.async_schedule_update_ha_state(True)
```
There's no `update` method of the entity in use, so remove argument `True`.

MartinHjelmare commented on requirements_all.txt:
```diff
@@ -919,6 +919,11 @@ pyhiveapi==0.2.14
 pyhomematic==0.1.51
 
+# homeassistant.components.homeworks
+# homeassistant.components.binary_sensor.homeworks
```
Run `script/gen_requirements.py` again and commit the changes to update this file.

dubnom committed:
More requested changes
Removed HomeworksController, it wasn't needed.

MartinHjelmare reviewed (commented):
Almost done.

dubnom committed:
Removed stale code.

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
Can be merged when build passes.

dubnom commented:
Awesome!  Thanks for all the help.  I have 3 more integrations to do.  I’ll try to take all of my learnings and apply them first.
Thanks again - Mike

From: Martin Hjelmare <notifications@github.com>
Sent: Sunday, December 16, 2018 4:52 PM
To: home-assistant/home-assistant <home-assistant@noreply.github.com>
Cc: Michael Dubno <michael@dubno.com>; Author <author@noreply.github.com>
Subject: Re: [home-assistant/home-assistant] Add Lutron Homeworks component (#18311)


Can be merged when build passes.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_home-2Dassistant_home-2Dassistant_pull_18311-23issuecomment-2D447679278&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=JTOZfzqesCAN0cvc6PpGOx114HUhWjfGXZHFS4MW31s&s=dUnMhlBZ4lNrF9GE7bieLMSJkNiFqS6b7rZ4dHe8mRs&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AQtHio-2DGgf5O0990Crgc-5FjNJqpTGxJdOks5u5sBmgaJpZM4YS9US&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=KoQiQ7r8wKDmbdHM2pr8o-7kSqCsNnMNSB2-Y-q4jAk&m=JTOZfzqesCAN0cvc6PpGOx114HUhWjfGXZHFS4MW31s&s=vlPplI2H-gc26sgGsaAujtaxkdzB4Z6HtO1VrBTj-hE&e=>.



________________________________

This email has been scanned for spam and viruses by Proofpoint Essentials. Click here<https://us3.proofpointessentials.com/index01.php?mod_id=11&mod_option=logitem&mail_id=1544997096-fI4ngLDcM92U&r_address=michael%40dubno.com&report=1> to report this email as spam.

cgarwood commented:
Travis doesn't like one of the docstrings:
`./homeassistant/components/light/homeworks.py:48:1: D401 First line should be in imperative mood`

dubnom committed:
Imperative doc change.

dubnom commented:
I don’t think it supports it.

Sent from my iPhone

On Feb 2, 2019, at 6:27 PM, bla <notifications@github.com<mailto:notifications@github.com>> wrote:


Do this one support Homeworks Illumination from 2008?
the processors come with ethernet - so not need box to convert rs232 to ethernet?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/home-assistant/home-assistant/pull/18311#issuecomment-460008394>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AQtHiu6T4G5YDnZ9AGOpWHvg94k8ellPks5vJh7DgaJpZM4YS9US>.

githubbla commented:
> I don’t think it supports it.
> […](#)
> Sent from my iPhone On Feb 2, 2019, at 6:27 PM, bla <notifications@github.com<mailto:notifications@github.com>> wrote: Do this one support Homeworks Illumination from 2008? the processors come with ethernet - so not need box to convert rs232 to ethernet? — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub<[#18311 (comment)](https://github.com/home-assistant/home-assistant/pull/18311#issuecomment-460008394)>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AQtHiu6T4G5YDnZ9AGOpWHvg94k8ellPks5vJh7DgaJpZM4YS9US>.
so to try it with rs232 to tcp/ip converter?

MartinHjelmare commented:
If you need help please use our help channels:
https://home-assistant.io/help/#communication-channels

Merged PRs should not be used for support. Thanks!
